
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>VCF &#8212; GenotypeFilesConvertor 1.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="BED-BIM-FAM" href="BED-BIM-FAM.html" />
    <link rel="prev" title="Welcome to GenotypeFilesConvertor’s documentation!" href="index.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="vcf">
<h1>VCF<a class="headerlink" href="#vcf" title="Permalink to this headline">¶</a></h1>
<div class="section" id="vcf-to-ped-map">
<h2>VCF to PED-MAP<a class="headerlink" href="#vcf-to-ped-map" title="Permalink to this headline">¶</a></h2>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">#</span> Include extension of file. For example, input_file.vcf.
<span class="go">./plink --vcf input_file.vcf --recode --out output_file</span>
</pre></div>
</div>
</div>
<div class="section" id="vcf-to-raw">
<h2>VCF to RAW<a class="headerlink" href="#vcf-to-raw" title="Permalink to this headline">¶</a></h2>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">#</span> Include extension of file. For example, input_file.vcf.
<span class="go">./plink --vcf input_file.vcf --recodeA --out output_file</span>
</pre></div>
</div>
</div>
<div class="section" id="vcf-to-bed-bim-fam">
<h2>VCF to BED-BIM-FAM<a class="headerlink" href="#vcf-to-bed-bim-fam" title="Permalink to this headline">¶</a></h2>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">#</span> Include extension of file. For example, input_file.vcf.
<span class="go">./plink --vcf input_file.vcf --make-bed --out output_file</span>
</pre></div>
</div>
</div>
<div class="section" id="vcf-to-gen-sample">
<h2>VCF to GEN-SAMPLE<a class="headerlink" href="#vcf-to-gen-sample" title="Permalink to this headline">¶</a></h2>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">#</span> Include extension of file. For example, input_file.vcf.
<span class="go">./plink --vcf input_file.vcf  --export oxford --out output_file</span>
</pre></div>
</div>
</div>
<div class="section" id="vcf-to-23andme">
<h2>VCF to 23andme<a class="headerlink" href="#vcf-to-23andme" title="Permalink to this headline">¶</a></h2>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">#</span> Input file should not include extension
<span class="gp">#</span> Make a directory in which 23andme files will be saved.
<span class="go">if not os.path.isdir(&quot;23andme&quot;):</span>
<span class="go">   os.mkdir(&quot;23andme&quot;)</span>

<span class="gp">#</span>Convert VCF to BED_BIM_FAM --&gt; VCFtoBED_BIM_FAM<span class="o">()</span>
<span class="go">./plink --vcf input_file+&quot;.vcf&quot; --make-bed --out output_file</span>

<span class="gp">#</span> It will generate three files output_file.bed, output_file.fam, and output_file.bim

<span class="gp">#</span>Extract id of each person
<span class="go">os.system(&quot;bcftools query -l &quot;+input_file+&quot; &gt; ./23andme/temp_samples.txt&quot;)</span>

<span class="gp">#</span>Open that file
<span class="go">f = open(&quot;./23andme/temp_samples.txt&quot;, &quot;r&quot;)</span>
<span class="go">for x in f:</span>
<span class="gp">   #</span>Write each person name in a specific file
<span class="go">   temp = open(&quot;./23andme/temp.txt&quot;, &quot;w&quot;)</span>

<span class="go">   temp.write(x.strip(&#39;\n&#39;).split(&quot;_&quot;)[0] +&quot;  &quot;+x.strip(&#39;\n&#39;).split(&quot;_&quot;)[1])</span>
<span class="go">   temp.close()</span>

<span class="gp">   #</span>Extract each person from BED,BIM,FAM file and convert it to 23andme.
<span class="go">   os.system(&quot;./plink --bfile &quot;+input_file.split(&quot;.&quot;)[0]+&quot; --keep ./23andme/temp.txt --recode 23 --snps-only --out ./23andme/&quot;+x.strip(&#39;\n&#39;))</span>
</pre></div>
</div>
</div>
<div class="section" id="vcf-to-ancestrydna">
<h2>VCF to AncestryDNA<a class="headerlink" href="#vcf-to-ancestrydna" title="Permalink to this headline">¶</a></h2>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">#</span>Input file should not include extension
<span class="go">Convert VCF to 23andme --&gt; VCFto23andme()</span>

<span class="gp">#</span>Convert 23andme to AncestryDNA --&gt;23andmeytoAncestryDNA<span class="o">()</span>
<span class="gp">#</span> Make a directory in which AncestryDNA files will be saved.
<span class="go">if not os.path.isdir(&quot;AncestryDNA&quot;):</span>
<span class="go">   os.mkdir(&quot;AncestryDNA&quot;)</span>
<span class="gp">  #</span>VCFto_23andme<span class="o">(</span>input_file<span class="o">)</span>

<span class="gp">#</span>Read 23andme files
<span class="go">_23andmefiles  = os.listdir(&#39;./23andme&#39;)</span>

<span class="gp">#</span>Read files one-by-one
<span class="go">for files in _23andmefiles:</span>
<span class="gp">   #</span> 23andme files are in .txt file format
<span class="go">   if &quot;.txt&quot; in files and &quot;temp&quot; not in files:</span>

<span class="gp">      #</span>Check size
<span class="go">      if os.stat(&quot;./23andme&quot;+os.sep+files).st_size == 0:</span>
<span class="go">         continue</span>
<span class="go">      else:</span>
<span class="go">         data = pd.read_csv(&quot;./23andme&quot;+os.sep+files,sep=&quot;\t&quot;,skiprows=8)</span>
<span class="go">         new = pd.DataFrame()</span>


<span class="go">         new[&#39;Rsid&#39;] = data[&#39;# rsid&#39;].values</span>
<span class="go">         new[&#39;Chromosome&#39;] = data[&#39;chromosome&#39;].values</span>
<span class="go">         new[&#39;position&#39;] = data[&#39;position&#39;].values</span>

<span class="gp">         #</span>Split genotype into allele1 and allele2
<span class="go">         new[&#39;allele1&#39;] = data[&#39;genotype&#39;].str[0]</span>
<span class="go">         new[&#39;allele2&#39;] =data[&#39;genotype&#39;].str[1]</span>

<span class="gp">         #</span>Change chromosome numbers
<span class="go">         new[&#39;Chromosome&#39;] = new[&#39;Chromosome&#39;].replace(23, &#39;X&#39;)</span>
<span class="go">         new[&#39;Chromosome&#39;] = new[&#39;Chromosome&#39;].replace(24, &#39;Y&#39;)</span>
<span class="go">         new[&#39;Chromosome&#39;] = new[&#39;Chromosome&#39;].replace(25, &#39;XY&#39;)</span>
<span class="go">         new[&#39;Chromosome&#39;] = new[&#39;Chromosome&#39;].replace(26, &#39;MT&#39;)</span>
<span class="go">         new.to_csv(&quot;./AncestryDNA&quot;+os.sep+files, sep=&quot;\t&quot;)</span>
</pre></div>
</div>
</div>
<div class="section" id="vcf-to-haps-legend-sample">
<h2>VCF to HAPS-LEGEND-SAMPLE<a class="headerlink" href="#vcf-to-haps-legend-sample" title="Permalink to this headline">¶</a></h2>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">bcftools convert input_file.vcf  -h  output_file</span>
</pre></div>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">GenotypeFilesConvertor</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">VCF</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#vcf-to-ped-map">VCF to PED-MAP</a></li>
<li class="toctree-l2"><a class="reference internal" href="#vcf-to-raw">VCF to RAW</a></li>
<li class="toctree-l2"><a class="reference internal" href="#vcf-to-bed-bim-fam">VCF to BED-BIM-FAM</a></li>
<li class="toctree-l2"><a class="reference internal" href="#vcf-to-gen-sample">VCF to GEN-SAMPLE</a></li>
<li class="toctree-l2"><a class="reference internal" href="#vcf-to-23andme">VCF to 23andme</a></li>
<li class="toctree-l2"><a class="reference internal" href="#vcf-to-ancestrydna">VCF to AncestryDNA</a></li>
<li class="toctree-l2"><a class="reference internal" href="#vcf-to-haps-legend-sample">VCF to HAPS-LEGEND-SAMPLE</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="BED-BIM-FAM.html">BED-BIM-FAM</a></li>
<li class="toctree-l1"><a class="reference internal" href="PED-MAP.html">PED-MAP</a></li>
<li class="toctree-l1"><a class="reference internal" href="GEN-SAMPLE.html">GEN-SAMPLE</a></li>
<li class="toctree-l1"><a class="reference internal" href="23andme.html">23andme</a></li>
<li class="toctree-l1"><a class="reference internal" href="HAPS-LEGEND-SAMPLE.html">HAPS-LEGEND-SAMPLE</a></li>
<li class="toctree-l1"><a class="reference internal" href="RAW.html">RAW</a></li>
<li class="toctree-l1"><a class="reference internal" href="AncestryDNA.html">AncestryDNA</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="index.html" title="previous chapter">Welcome to GenotypeFilesConvertor’s documentation!</a></li>
      <li>Next: <a href="BED-BIM-FAM.html" title="next chapter">BED-BIM-FAM</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021, Muhammad Muneeb.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/VCF.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>